clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -194,7 +107,27 @@ private ProduceRequest build(short version, boolean validate) {
                     ProduceRequest.validateRecords(version, records);
                 }
             }
-            return new ProduceRequest(version, acks, timeout, partitionRecords, transactionalId);
+
+            List<ProduceRequestData.TopicProduceData> tpd = partitionRecords
```

Comment:
I wonder if we would get any benefit computing `partitionSizes` during this pass.

Reply:
We could probably also compute `partitionSizes` lazily. I think the broker is the only one that uses it.

Reply:
The "data" may be cleared to null (for GC) so it is a risk if we lazily initialize the partition size.

I considered to remove it as it is used by log only (toString).

Reply:
Right. I thought we might consider computing it just before the data gets cleared.